src/main/java/org/embulk/util/json/InternalJsonValuesReader.java

```diff
@@ -0,0 +1,440 @@
+/*
+ * Copyright 2023 The Embulk project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.embulk.util.json;
+
+import com.fasterxml.jackson.core.JsonParser;
+import com.fasterxml.jackson.core.JsonToken;
+import java.io.IOException;
+import java.util.AbstractMap;
+import java.util.ArrayDeque;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+import org.embulk.spi.json.JsonArray;
+import org.embulk.spi.json.JsonBoolean;
+import org.embulk.spi.json.JsonDouble;
+import org.embulk.spi.json.JsonLong;
+import org.embulk.spi.json.JsonNull;
+import org.embulk.spi.json.JsonObject;
+import org.embulk.spi.json.JsonString;
+import org.embulk.spi.json.JsonValue;
+
+class InternalJsonValuesReader {
+    InternalJsonValuesReader(
+            final JsonParser parser,
+            final JsonPointerTree tree,
+            final int size,
+            final boolean withNumbersFallbackWithLiterals,
+            final double defaultDouble,
+            final long defaultLong) {
+        this.parser = parser;
+        this.tree = tree;
+
+        this.withNumbersFallbackWithLiterals = withNumbersFallbackWithLiterals;
+        this.defaultDouble = defaultDouble;
+        this.defaultLong = defaultLong;
+        this.size = size;
+
+        this.pointerStack = new ArrayDeque<>();
```

Comment:
Just question, Are these stack `LIFO`, 'FIFO` or something?

Reply:
Intends to be LIFO. `ArrayDeque` is designed to be used for both LIFO and FIFO.

https://docs.oracle.com/javase/8/docs/api/java/util/ArrayDeque.html

Reply:
It meant If I knew `parsingStack` is designed for `LIFO`, It thought why this stack get the first entry like `FIFO`?

```java
final ParsingContext context = this.parsingStack.getFirst(); // getLast?
```

Reply:
As our conversation outside, `push`/`pop` (like LIFO stack) for `ArrayDeque` works for the first, not the last.